### PR TITLE
[Protocol3] Batched proof verification

### PR DESCRIPTION
**Pull request by Brechtpd**
Reduces the cost of verifying a proof from 550,000 gas/proof to about 200,000 gas/proof.

### CODE REVIEW

**Comment by Brechtpd** [comment]
@dong77 Ready for review.

Unfortunately this PR has an extremely messy diff because of the automatic formatting that is now done when committing.

**Comment by dong77** [comment]
> @dong77 Ready for review.
> 
> Unfortunately this PR has an extremely messy diff because of the automatic formatting that is now done when committing.

@Brechtpd no worries. @letsgoustc and myself will probably need a week to review this PR. Please be patient.

File: packages/loopring_v3/contracts/impl/BlockVerifier.sol [hunk 1]
```diff
@@ -41,54 +48,92 @@ contract BlockVerifier is IBlockVerifier, Claimable
         external
         onlyOwner
     {
-        // While vk[0] could be 0, we don't allow it so we can use this to easily check
-        // if the verification key is set
-        require(vk[0] != 0, "INVALID_DATA");
         bool dataAvailability = needsDataAvailability(blockType, onchainDataAvailability);
         require(dataAvailability == onchainDataAvailability, "NO_DATA_AVAILABILITY_NEEDED");
+        Circuit storage circuit = circuits[onchainDataAvailability][blockType][blockSize][blockVersion];
+        require(circuit.registered == false, "ALREADY_REGISTERED");
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/BlockVerifier.sol]: It seems:
1) once a circuit is disabled, it cannot be re-enabled. 
2) once a circuit is registered, we cannot re-register it again with a new set of new keys. 

Both seem OK to me.

Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/BlockVerifier.sol]: That's correct.

File: packages/loopring_v3/contracts/impl/BlockVerifier.sol [hunk 2]
```diff
@@ -41,54 +48,92 @@ contract BlockVerifier is IBlockVerifier, Claimable
...
+        onlyOwner
     {
-        bool dataAvailability = needsDataAvailability(blockType, onchainDataAvailability);
-        return verificationKeys[dataAvailability][blockType][blockSize][blockVersion][0] != 0;
+        Circuit storage circuit = circuits[onchainDataAvailability][blockType][blockSize][blockVersion];
+        require(circuit.registered == true, "NOT_REGISTERED");
+        require(circuit.enabled == true, "ALREADY_DISABLED");
+        circuit.enabled = false;
+
+        emit CircuitDisabled(blockType, onchainDataAvailability, blockSize, blockVersion);
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/BlockVerifier.sol]: break into multiple lines.

File: packages/loopring_v3/contracts/impl/BlockVerifier.sol [hunk 3]
```diff
@@ -41,54 +48,92 @@ contract BlockVerifier is IBlockVerifier, Claimable
...
+        Circuit storage circuit = circuits[onchainDataAvailability][blockType][blockSize][blockVersion];
+        require(circuit.registered == false, "ALREADY_REGISTERED");
         for (uint i = 0; i < 18; i++) {
-            verificationKeys[onchainDataAvailability][blockType][blockSize][blockVersion][i] = vk[i];
+            circuit.verificationKey[i] = vk[i];
         }
+        circuit.registered = true;
+        circuit.enabled = true;
+
+        emit CircuitRegistered(blockType, onchainDataAvailability, blockSize, blockVersion);
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/BlockVerifier.sol]: break into multiple lines.

File: packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol [hunk 4]
```diff
@@ -82,52 +82,88 @@ library ExchangeBlocks
         );
     }
 
-    function verifyBlock(
+    function verifyBlocks(
         ExchangeData.State storage S,
-        uint blockIdx,
-        uint256[8] memory proof
+        uint[] memory blockIndices,
```
Comment (dong77) [inline packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol]: should we check the max length of blockIndies?

Comment (Brechtpd) [reply packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol]: There is no hard limit (except for the Ethereum gas limit/block). There's also a good chance verifying a proof will get even cheaper in Istanbul so it's up to the operator to decide how many proofs can be successfully verified.

**Pull request merged**